File: backend/app/routes/quizzes.py
from uuid import UUID

# pyrefly: ignore [missing-import]
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.auth.dependencies import get_current_user, require_student, require_teacher
from app.db.session import get_db
from app.models.quiz import Quiz
from app.models.user import User
from app.schemas.common import StandardResponse
from app.schemas.quiz import (
    QuizAttemptCreateRequest,
    QuizAttemptResponse,
    QuizCreateRequest,
    QuizGenerateRequest,
    QuizGenerateResponse,
    QuizQuestionOut,
    QuizQuestionStudentOut,
    QuizStudentDetailOut,
    QuizSummaryOut,
    QuizTeacherDetailOut,
    QuizUpdateRequest,
)
from app.services.quiz.ai_generate_service import AIQuizGenerateService
from app.services.quiz.grading_service import (
    get_own_attempt,
    serialize_attempt,
    submit_attempt,
)
from app.services.quiz.manual_service import (
    create_draft,
    delete_quiz,
    get_student_quiz,
    get_teacher_quiz,
    list_student_quizzes,
    list_teacher_quizzes,
    publish_quiz,
    update_draft,
)

router = APIRouter(prefix="/api", tags=["Quizzes"])


# ── Serialization helpers ─────────────────────────────────────────────────────

def _summary_out(quiz: Quiz) -> QuizSummaryOut:
    return QuizSummaryOut(
        id=quiz.id,
        subject_id=quiz.subject_id,
        teacher_id=quiz.teacher_id,
        teacher_name=quiz.teacher.name if quiz.teacher else None,
        topic=quiz.topic,
        source=quiz.source,
        status=quiz.status,
        time_limit_seconds=quiz.time_limit_seconds,
        question_count=len(quiz.questions),
        created_at=quiz.created_at,
    )


def _teacher_detail_out(quiz: Quiz) -> QuizTeacherDetailOut:
    return QuizTeacherDetailOut(
        id=quiz.id,
        subject_id=quiz.subject_id,
        teacher_id=quiz.teacher_id,
        teacher_name=quiz.teacher.name if quiz.teacher else None,
        topic=quiz.topic,
        source=quiz.source,
        status=quiz.status,
        time_limit_seconds=quiz.time_limit_seconds,
        created_at=quiz.created_at,
        questions=[QuizQuestionOut.model_validate(q) for q in quiz.questions],
    )


def _student_detail_out(quiz: Quiz) -> QuizStudentDetailOut:
    """Student-facing detail. Uses QuizQuestionStudentOut exclusively so
    `correct_option` is never serialised."""
    return QuizStudentDetailOut(
        id=quiz.id,
        subject_id=quiz.subject_id,
        teacher_id=quiz.teacher_id,
        teacher_name=quiz.teacher.name if quiz.teacher else None,
        topic=quiz.topic,
        source=quiz.source,
        status=quiz.status,
        time_limit_seconds=quiz.time_limit_seconds,
        created_at=quiz.created_at,
        questions=[QuizQuestionStudentOut.model_validate(q) for q in quiz.questions],
    )


def _attempt_out(attempt) -> QuizAttemptResponse:
    return serialize_attempt(attempt)


# ── Quiz list / detail (role-aware) ───────────────────────────────────────────

@router.get("/quizzes", response_model=StandardResponse)
def list_quizzes(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Teacher: all quizzes (draft + published) in subjects they teach.
    Student: published quizzes only, in enrolled subjects."""
    if current_user.role == "teacher":
        quizzes = list_teacher_quizzes(db, current_user)
    elif current_user.role == "student":
        quizzes = list_student_quizzes(db, current_user)
    else:
        raise HTTPException(status_code=403, detail="Unknown role")
    return StandardResponse.ok(data=[_summary_out(q).model_dump(mode="json") for q in quizzes])


@router.get("/quizzes/{quiz_id}", response_model=StandardResponse)
def get_quiz(
    quiz_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Teacher detail includes correct_option; student detail never does."""
    if current_user.role == "teacher":
        quiz = get_teacher_quiz(db, current_user, quiz_id)
        data = _teacher_detail_out(quiz)
    elif current_user.role == "student":
        quiz = get_student_quiz(db, current_user, quiz_id)
        data = _student_detail_out(quiz)
    else:
        raise HTTPException(status_code=403, detail="Unknown role")
    return StandardResponse.ok(data=data.model_dump(mode="json"))


# ── Teacher authoring ─────────────────────────────────────────────────────────

@router.post("/quizzes", response_model=StandardResponse, status_code=201)
def create_quiz(
    request: QuizCreateRequest,
    current_user: User = Depends(require_teacher),
    db: Session = Depends(get_db),
):
    quiz = create_draft(db, current_user, request)
    return StandardResponse.ok(data=_teacher_detail_out(quiz).model_dump(mode="json"))


@router.patch("/quizzes/{quiz_id}", response_model=StandardResponse)
def update_quiz(
    quiz_id: UUID,
    request: QuizUpdateRequest,
    current_user: User = Depends(require_teacher),
    db: Session = Depends(get_db),
):
    quiz = update_draft(db, current_user, quiz_id, request)
    return StandardResponse.ok(data=_teacher_detail_out(quiz).model_dump(mode="json"))


@router.delete("/quizzes/{quiz_id}", response_model=StandardResponse)
def remove_quiz(
    quiz_id: UUID,
    current_user: User = Depends(require_teacher),
    db: Session = Depends(get_db),
):
    delete_quiz(db, current_user, quiz_id)
    return StandardResponse.ok(data={"deleted": True})


@router.post("/quizzes/{quiz_id}/publish", response_model=StandardResponse)
def publish_quiz_route(
    quiz_id: UUID,
    current_user: User = Depends(require_teacher),
    db: Session = Depends(get_db),
):
    quiz = publish_quiz(db, current_user, quiz_id)
    return StandardResponse.ok(data=_teacher_detail_out(quiz).model_dump(mode="json"))


@router.post("/quiz/generate", response_model=StandardResponse)
def generate_quiz_draft(
    request: QuizGenerateRequest,
    current_user: User = Depends(require_teacher),
    db: Session = Depends(get_db),
):
    """Generate a draft question set from ready materials for review. Nothing
    is inserted — the teacher must save it as a quiz before publishing."""
    questions = AIQuizGenerateService().generate_draft(db, current_user, request)
    response = QuizGenerateResponse(questions=questions)
    return StandardResponse.ok(data=response.model_dump(mode="json"))


# ── Student attempts ──────────────────────────────────────────────────────────

@router.post("/quiz-attempts", response_model=StandardResponse)
def create_attempt(
    request: QuizAttemptCreateRequest,
    current_user: User = Depends(require_student),
    db: Session = Depends(get_db),
):
    """Submit answers. Idempotent: retries return the existing attempt (200)
    with the same result rather than creating a duplicate row."""
    attempt = submit_attempt(
        db,
        current_user,
        request.quiz_id,
        {str(question_id): option for question_id, option in request.answers.items()},
    )
    return StandardResponse.ok(data=_attempt_out(attempt).model_dump(mode="json"))


@router.get("/quiz-attempts/{attempt_id}", response_model=StandardResponse)
def get_attempt(
    attempt_id: UUID,
    current_user: User = Depends(require_student),
    db: Session = Depends(get_db),
):
    attempt = get_own_attempt(db, current_user, attempt_id)
    return StandardResponse.ok(data=_attempt_out(attempt).model_dump(mode="json"))

File: frontend/src/api/quizzes.js
import { request } from './client';

// ── Quiz summaries / detail ────────────────────────────────────────────────

export function mapQuizSummary(q) {
  return {
    id: q.id,
    subjectId: q.subject_id,
    teacherId: q.teacher_id,
    teacherName: q.teacher_name ?? null,
    topic: q.topic,
    source: q.source,
    status: q.status,
    timeLimitSeconds: q.time_limit_seconds ?? null,
    questionCount: q.question_count ?? 0,
    createdAt: q.created_at,
  };
}

/** GET /api/quizzes — role-aware: teacher sees drafts+published, student published only. */
export async function listQuizzes() {
  const data = await request('/api/quizzes');
  return (data || []).map(mapQuizSummary);
}

/** GET /api/quizzes/:id — role-aware; student variant never leaks correct answers. */
export async function getQuiz(id) {
  const data = await request(`/api/quizzes/${id}`);
  return {
    id: data.id,
    subjectId: data.subject_id,
    teacherId: data.teacher_id,
    teacherName: data.teacher_name ?? null,
    topic: data.topic,
    source: data.source,
    status: data.status,
    timeLimitSeconds: data.time_limit_seconds ?? null,
    createdAt: data.created_at,
    questions: (data.questions || []).map((q) => ({
      id: q.id,
      stem: q.question_text,
      options: q.options,
      correct: q.correct_option ?? null,
      topicTag: q.topic_tag ?? null,
      difficulty: q.difficulty ?? null,
    })),
  };
}

// ── Teacher authoring ──────────────────────────────────────────────────────

function toQuestionInput(q) {
  return {
    question_text: q.stem,
    options: q.options,
    correct_option: q.correct,
    topic_tag: q.topicTag ?? null,
    difficulty: q.difficulty ?? 'medium',
  };
}

/** POST /api/quizzes — creates a draft. */
export async function createQuiz({ subjectId, topic, source = 'manual', timeLimitSeconds, questions }) {
  const data = await request('/api/quizzes', {
    method: 'POST',
    body: {
      subject_id: subjectId,
      topic,
      source,
      time_limit_seconds: timeLimitSeconds ?? null,
      questions: questions.map(toQuestionInput),
    },
  });
  return data.id;
}

/** PATCH /api/quizzes/:id — update a draft in place. */
export async function updateQuiz(id, { topic, timeLimitSeconds, questions } = {}) {
  const body = {};
  if (topic !== undefined) body.topic = topic;
  if (timeLimitSeconds !== undefined) body.time_limit_seconds = timeLimitSeconds;
  if (questions !== undefined) body.questions = questions.map(toQuestionInput);
  return request(`/api/quizzes/${id}`, { method: 'PATCH', body });
}

/** DELETE /api/quizzes/:id */
export async function deleteQuiz(id) {
  return request(`/api/quizzes/${id}`, { method: 'DELETE' });
}

/** POST /api/quizzes/:id/publish */
export async function publishQuiz(id) {
  return request(`/api/quizzes/${id}/publish`, { method: 'POST' });
}

/** POST /api/quiz/generate — AI draft, not persisted. */
export async function generateQuiz({ subjectId, materialIds, topic, questionCount = 10 }) {
  const data = await request('/api/quiz/generate', {
    method: 'POST',
    body: {
      subject_id: subjectId,
      material_ids: materialIds,
      topic,
      question_count: questionCount,
    },
  });
  return (data.questions || []).map((q) => ({
    stem: q.question_text,
    options: q.options,
    correct: q.correct_option,
    topicTag: q.topic_tag ?? null,
    difficulty: q.difficulty ?? 'medium',
  }));
}

// ── Student attempts ───────────────────────────────────────────────────────

function mapAttempt(a) {
  return {
    id: a.id,
    quizId: a.quiz_id,
    studentId: a.student_id,
    quizTitle: a.quiz_title,
    subject: a.subject_name ?? null,
    answers: a.answers || {},
    score: a.score,
    correctCount: a.correct_count,
    totalQuestions: a.total_questions,
    weakTopics: (a.weak_topics || []).map((wt) => ({ topic: wt.topic, accuracy: wt.accuracy })),
    submittedAt: a.submitted_at,
    questions: (a.questions || []).map((q) => ({
      id: q.question_id,
      stem: q.question_text,
      options: q.options,
      correct: q.correct_option,
      selected: q.selected_option ?? null,
      isCorrect: q.is_correct,
    })),
  };
}

/**
 * POST /api/quiz-attempts — submit answers.
 * `answers` maps questionId → option TEXT (the backend grades option strings).
 * Idempotent: resubmitting returns the existing attempt.
 */
export async function submitAttempt({ quizId, answers }) {
  const data = await request('/api/quiz-attempts', {
    method: 'POST',
    body: { quiz_id: quizId, answers },
  });
  return mapAttempt(data);
}

/** GET /api/quiz-attempts/:id — a single own attempt. */
export async function getAttempt(id) {
  const data = await request(`/api/quiz-attempts/${id}`);
  return mapAttempt(data);
}

/** GET /api/students/me/attempts — own attempts (newest first), optional quiz filter. */
export async function listMyAttempts({ quizId } = {}) {
  const data = await request('/api/students/me/attempts', {
    params: quizId ? { quiz_id: quizId } : undefined,
  });
  return (data || []).map(mapAttempt);
}
